macros/contexts/contextSignificantFigures.pl

```diff
@@ -136,134 +192,150 @@ sub new {
+# Return the exponent.
+sub E { shift->{exp} }
 
-#  Stringify and TeXify the number in the context's base
-sub string {
-	my $self = shift;
-	return "" . $self->value if $self->sigfigs =~ /inf/;
-	my $N = -$self->{exp} + $self->sigfigs - 1;     # digit to round to.
-	my $v = $self->mantissa * 10**($self->{exp});
-	return abs($self->{exp}) > 5 ? $self->value : ($N > 0 ? sprintf("%.${N}f", $v) : "$v");
+# Return the exponential for the given $value with max($n,14-$n) sigfigs.  This is
+# helpful for addition and subtraction.
+sub expFor {
+	my ($self, $value, $n) = @_;
+	$n = main::max($n, 14 - $n);
+	return (split(/E/, $self->format('E', $value, $n)))[1] + 0;
 }
 
+#  Stringify and TeXify the number in the context's base
+sub string { shift->format('f') }
+
 sub TeX {
-	my $self = shift;
-	my $tex  = $self->string;
+	my $tex = shift->string;
 	$tex =~ s/E(?:(-)|\+)0*([1-9]\d?)/\\times 10^{$1$2}/;
 	return "{$tex}";
 }
 
-sub format {
-	my ($self, $char, $value, $n) = @_;
-	my $m = $n - 1;
-	return sprintf("%.${m}E", $value + 0);
+# Determine the exponent of the given number.  Works like log10, but handles 0 and negatives.
+
+sub expon {
+	my $value = sprintf("%E", shift);
+	return ((split(/E/, $value))[1]) + 0;
 }
 
-# Redefine adding two numbers.  This calculates the last significant digit and then converts each number
-# to an integer with the ones being signficant.  Then add and round, finally converting to decimal using the
-# last significant digit as the exponent.
+# Format the number in $value in either 'E' (exponential form) or 'f' decimal form using
+# $n signficant figures.
+
+# Example: format('E', '123.456', 6) returns 1.23456E+02
+# format('f', 1.23E-01, 3) returns '0.123'.
+
+sub format {
+	my $self = shift;
+	my ($f, $value, $n);
+	($self, $f, $value, $n) = ref($self) eq 'context::SignificantFigures::Real' ? ($self, @_) : (undef, $self, @_);
+	$value = $self->value unless defined $value;
+	$n     = $self->N     unless defined $n;
+	return "$value" if $n == 'inf';
+
+	my $exp = defined($self) ? $self->E : expon($value);
+	$f = 'E' if $f eq 'f' && ($n < 1 || $exp >= 5 || -5 >= $exp);
+	$n -= $exp if $f eq 'f';
+	$n = main::max(0, $n - 1);
+	return sprintf("%.${n}${f}" . ($n == 0 && $f eq 'f' ? '.' : ''), $value);
+}
 
-# If the first addend is negative, write it as a subtraction.
+# Redefine addition.  The leftmost signifcant place in the result is needed to get the
+# correct value.
 
 sub add {
 	my ($self, $l, $r, $other) = Value::checkOpOrderWithPromote(@_);
-	return $r - abs($l) if $l->value < 0;
-	my $value = $l->value + $r->value;
-	my $exp   = int(log($value) / log(10));
-	my $n     = main::min($l->{sigfigs} - $l->{exp}, $r->{sigfigs} - $r->{exp}) + $exp;
-	return $n <= 0 ? $self->new(0) : $self->new($value, sigfigs => $n);
+	my $exp   = main::min($l->N - $l->E, $r->N - $r->E);
+	my $value = $l->round($exp) + $r->round($exp);
+	return $self->new($value, sigfigs => $exp + $self->expFor($value, $exp));
```

Comment:
In doing some more testing, it looks like you need to restrict the `sigfigs` using

``` perl
	return $self->new($value, sigfigs => main::max(0, $exp + $self->expFor($value, $exp)));
```

otherwise `Real(-100.005 + 100)` will report an error about negative significant figures.